File: swift4/magic-the-gathering.swift
// To parse the JSON, add this file to your project and do:
//
//   let lEAExtras = try LEAExtras(json)

import Foundation

struct LEAExtras: Codable {
    let name, code, gathererCode, magicCardsInfoCode: String
    let releaseDate, border, type: String
    let booster: [Booster]
    let mkmName: String
    let mkmID: Int
    let cards: [Card]

    enum CodingKeys: String, CodingKey {
        case name, code, gathererCode, magicCardsInfoCode, releaseDate, border, type, booster
        case mkmName = "mkm_name"
        case mkmID = "mkm_id"
        case cards
    }
}

enum Booster: String, Codable {
    case common = "common"
    case rare = "rare"
    case uncommon = "uncommon"
}

struct Card: Codable {
    let artist: String
    let cmc: Int
    let id, imageName: String
    let layout: Layout
    let legalities: [LegalityElement]
    let manaCost, mciNumber: String?
    let multiverseid: Int
    let name: String
    let originalText: String?
    let originalType: String
    let printings: [String]
    let rarity: Rarity
    let rulings: [Ruling]?
    let text: String?
    let type: String
    let types: [TypeElement]
    let reserved: Bool?
    let power: String?
    let subtypes: [String]?
    let toughness: String?
    let colorIdentity: [ColorIdentity]?
    let flavor: String?
    let colors: [Color]?
    let supertypes: [Supertype]?
    let variations: [Int]?
}

enum ColorIdentity: String, Codable {
    case b = "B"
    case g = "G"
    case r = "R"
    case u = "U"
    case w = "W"
}

enum Color: String, Codable {
    case black = "Black"
    case blue = "Blue"
    case green = "Green"
    case red = "Red"
    case white = "White"
}

enum Layout: String, Codable {
    case normal = "normal"
}

struct LegalityElement: Codable {
    let format: Format
    let legality: LegalityEnum
}

enum Format: String, Codable {
    case brawl = "Brawl"
    case commander = "Commander"
    case legacy = "Legacy"
    case modern = "Modern"
    case standard = "Standard"
    case vintage = "Vintage"
}

enum LegalityEnum: String, Codable {
    case banned = "Banned"
    case legal = "Legal"
    case restricted = "Restricted"
}

enum Rarity: String, Codable {
    case basicLand = "Basic Land"
    case common = "Common"
    case rare = "Rare"
    case uncommon = "Uncommon"
}

struct Ruling: Codable {
    let date, text: String
}

enum Supertype: String, Codable {
    case basic = "Basic"
}

enum TypeElement: String, Codable {
    case artifact = "Artifact"
    case creature = "Creature"
    case enchantment = "Enchantment"
    case instant = "Instant"
    case land = "Land"
    case sorcery = "Sorcery"
}

// MARK: Convenience initializers and mutators

extension LEAExtras {
    init(data: Data) throws {
        self = try newJSONDecoder().decode(LEAExtras.self, from: data)
    }

    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
        guard let data = json.data(using: encoding) else {
            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
        }
        try self.init(data: data)
    }

    init(fromURL url: URL) throws {
        try self.init(data: try Data(contentsOf: url))
    }

    func with(
        name: String? = nil,
        code: String? = nil,
        gathererCode: String? = nil,
        magicCardsInfoCode: String? = nil,
        releaseDate: String? = nil,
        border: String? = nil,
        type: String? = nil,
        booster: [Booster]? = nil,
        mkmName: String? = nil,
        mkmID: Int? = nil,
        cards: [Card]? = nil
    ) -> LEAExtras {
        return LEAExtras(
            name: name ?? self.name,
            code: code ?? self.code,
            gathererCode: gathererCode ?? self.gathererCode,
            magicCardsInfoCode: magicCardsInfoCode ?? self.magicCardsInfoCode,
            releaseDate: releaseDate ?? self.releaseDate,
            border: border ?? self.border,
            type: type ?? self.type,
            booster: booster ?? self.booster,
            mkmName: mkmName ?? self.mkmName,
            mkmID: mkmID ?? self.mkmID,
            cards: cards ?? self.cards
        )
    }

    func jsonData() throws -> Data {
        return try newJSONEncoder().encode(self)
    }

    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
        return String(data: try self.jsonData(), encoding: encoding)
    }
}

extension Card {
    init(data: Data) throws {
        self = try newJSONDecoder().decode(Card.self, from: data)
    }

    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
        guard let data = json.data(using: encoding) else {
            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
        }
        try self.init(data: data)
    }

    init(fromURL url: URL) throws {
        try self.init(data: try Data(contentsOf: url))
    }

    func with(
        artist: String? = nil,
        cmc: Int? = nil,
        id: String? = nil,
        imageName: String? = nil,
        layout: Layout? = nil,
        legalities: [LegalityElement]? = nil,
        manaCost: String?? = nil,
        mciNumber: String?? = nil,
        multiverseid: Int? = nil,
        name: String? = nil,
        originalText: String?? = nil,
        originalType: String? = nil,
        printings: [String]? = nil,
        rarity: Rarity? = nil,
        rulings: [Ruling]?? = nil,
        text: String?? = nil,
        type: String? = nil,
        types: [TypeElement]? = nil,
        reserved: Bool?? = nil,
        power: String?? = nil,
        subtypes: [String]?? = nil,
        toughness: String?? = nil,
        colorIdentity: [ColorIdentity]?? = nil,
        flavor: String?? = nil,
        colors: [Color]?? = nil,
        supertypes: [Supertype]?? = nil,
        variations: [Int]?? = nil
    ) -> Card {
        return Card(
            artist: artist ?? self.artist,
            cmc: cmc ?? self.cmc,
            id: id ?? self.id,
            imageName: imageName ?? self.imageName,
            layout: layout ?? self.layout,
            legalities: legalities ?? self.legalities,
            manaCost: manaCost ?? self.manaCost,
            mciNumber: mciNumber ?? self.mciNumber,
            multiverseid: multiverseid ?? self.multiverseid,
            name: name ?? self.name,
            originalText: originalText ?? self.originalText,
            originalType: originalType ?? self.originalType,
            printings: printings ?? self.printings,
            rarity: rarity ?? self.rarity,
            rulings: rulings ?? self.rulings,
            text: text ?? self.text,
            type: type ?? self.type,
            types: types ?? self.types,
            reserved: reserved ?? self.reserved,
            power: power ?? self.power,
            subtypes: subtypes ?? self.subtypes,
            toughness: toughness ?? self.toughness,
            colorIdentity: colorIdentity ?? self.colorIdentity,
            flavor: flavor ?? self.flavor,
            colors: colors ?? self.colors,
            supertypes: supertypes ?? self.supertypes,
            variations: variations ?? self.variations
        )
    }

    func jsonData() throws -> Data {
        return try newJSONEncoder().encode(self)
    }

    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
        return String(data: try self.jsonData(), encoding: encoding)
    }
}

extension LegalityElement {
    init(data: Data) throws {
        self = try newJSONDecoder().decode(LegalityElement.self, from: data)
    }

    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
        guard let data = json.data(using: encoding) else {
            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
        }
        try self.init(data: data)
    }

    init(fromURL url: URL) throws {
        try self.init(data: try Data(contentsOf: url))
    }

    func with(
        format: Format? = nil,
        legality: LegalityEnum? = nil
    ) -> LegalityElement {
        return LegalityElement(
            format: format ?? self.format,
            legality: legality ?? self.legality
        )
    }

    func jsonData() throws -> Data {
        return try newJSONEncoder().encode(self)
    }

    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
        return String(data: try self.jsonData(), encoding: encoding)
    }
}

extension Ruling {
    init(data: Data) throws {
        self = try newJSONDecoder().decode(Ruling.self, from: data)
    }

    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
        guard let data = json.data(using: encoding) else {
            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
        }
        try self.init(data: data)
    }

    init(fromURL url: URL) throws {
        try self.init(data: try Data(contentsOf: url))
    }

    func with(
        date: String? = nil,
        text: String? = nil
    ) -> Ruling {
        return Ruling(
            date: date ?? self.date,
            text: text ?? self.text
        )
    }

    func jsonData() throws -> Data {
        return try newJSONEncoder().encode(self)
    }

    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
        return String(data: try self.jsonData(), encoding: encoding)
    }
}

func newJSONDecoder() -> JSONDecoder {
    let decoder = JSONDecoder()
    if #available(iOS 10.0, OSX 10.12, tvOS 10.0, watchOS 3.0, *) {
        decoder.dateDecodingStrategy = .iso8601
    }
    return decoder
}

func newJSONEncoder() -> JSONEncoder {
    let encoder = JSONEncoder()
    if #available(iOS 10.0, OSX 10.12, tvOS 10.0, watchOS 3.0, *) {
        encoder.dateEncodingStrategy = .iso8601
    }
    return encoder
}

File: flow/magic-the-gathering.js
// To parse this data:
//
//   const Convert = require("./file");
//
//   const lEAExtras = Convert.toLEAExtras(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
function toLEAExtras(json) {
    return cast(JSON.parse(json), r("LEAExtras"));
}

function lEAExtrasToJson(value) {
    return JSON.stringify(uncast(value, r("LEAExtras")), null, 2);
}

function invalidValue(typ, val) {
    throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`);
}

function jsonToJSProps(typ) {
    if (typ.jsonToJS === undefined) {
        var map = {};
        typ.props.forEach((p) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ) {
    if (typ.jsToJSON === undefined) {
        var map = {};
        typ.props.forEach((p) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val, typ, getProps) {
    function transformPrimitive(typ, val) {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val);
    }

    function transformUnion(typs, val) {
        // val must validate against one typ in typs
        var l = typs.length;
        for (var i = 0; i < l; i++) {
            var typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val);
    }

    function transformEnum(cases, val) {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases, val);
    }

    function transformArray(typ, val) {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue("array", val);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformObject(props, additional, val) {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue("object", val);
        }
        var result = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val);
    }
    if (typ === false) return invalidValue(typ, val);
    while (typeof typ === "object" && typ.ref !== undefined) {
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val);
    }
    return transformPrimitive(typ, val);
}

function cast(val, typ) {
    return transform(val, typ, jsonToJSProps);
}

function uncast(val, typ) {
    return transform(val, typ, jsToJSONProps);
}

function a(typ) {
    return { arrayItems: typ };
}

function u(...typs) {
    return { unionMembers: typs };
}

function o(props, additional) {
    return { props, additional };
}

function m(additional) {
    return { props: [], additional };
}

function r(name) {
    return { ref: name };
}

const typeMap = {
    "LEAExtras": o([
        { json: "name", js: "name", typ: "" },
        { json: "code", js: "code", typ: "" },
        { json: "gathererCode", js: "gathererCode", typ: "" },
        { json: "magicCardsInfoCode", js: "magicCardsInfoCode", typ: "" },
        { json: "releaseDate", js: "releaseDate", typ: "" },
        { json: "border", js: "border", typ: "" },
        { json: "type", js: "type", typ: "" },
        { json: "booster", js: "booster", typ: a(r("Booster")) },
        { json: "mkm_name", js: "mkm_name", typ: "" },
        { json: "mkm_id", js: "mkm_id", typ: 0 },
        { json: "cards", js: "cards", typ: a(r("Card")) },
    ], false),
    "Card": o([
        { json: "artist", js: "artist", typ: "" },
        { json: "cmc", js: "cmc", typ: 0 },
        { json: "id", js: "id", typ: "" },
        { json: "imageName", js: "imageName", typ: "" },
        { json: "layout", js: "layout", typ: r("Layout") },
        { json: "legalities", js: "legalities", typ: a(r("LegalityElement")) },
        { json: "manaCost", js: "manaCost", typ: u(undefined, "") },
        { json: "mciNumber", js: "mciNumber", typ: u(undefined, "") },
        { json: "multiverseid", js: "multiverseid", typ: 0 },
        { json: "name", js: "name", typ: "" },
        { json: "originalText", js: "originalText", typ: u(undefined, "") },
        { json: "originalType", js: "originalType", typ: "" },
        { json: "printings", js: "printings", typ: a("") },
        { json: "rarity", js: "rarity", typ: r("Rarity") },
        { json: "rulings", js: "rulings", typ: u(undefined, a(r("Ruling"))) },
        { json: "text", js: "text", typ: u(undefined, "") },
        { json: "type", js: "type", typ: "" },
        { json: "types", js: "types", typ: a(r("Type")) },
        { json: "reserved", js: "reserved", typ: u(undefined, true) },
        { json: "power", js: "power", typ: u(undefined, "") },
        { json: "subtypes", js: "subtypes", typ: u(undefined, a("")) },
        { json: "toughness", js: "toughness", typ: u(undefined, "") },
        { json: "colorIdentity", js: "colorIdentity", typ: u(undefined, a(r("ColorIdentity"))) },
        { json: "flavor", js: "flavor", typ: u(undefined, "") },
        { json: "colors", js: "colors", typ: u(undefined, a(r("Color"))) },
        { json: "supertypes", js: "supertypes", typ: u(undefined, a(r("Supertype"))) },
        { json: "variations", js: "variations", typ: u(undefined, a(0)) },
    ], false),
    "LegalityElement": o([
        { json: "format", js: "format", typ: r("Format") },
        { json: "legality", js: "legality", typ: r("LegalityEnum") },
    ], false),
    "Ruling": o([
        { json: "date", js: "date", typ: "" },
        { json: "text", js: "text", typ: "" },
    ], false),
    "Booster": [
        "common",
        "rare",
        "uncommon",
    ],
    "ColorIdentity": [
        "B",
        "G",
        "R",
        "U",
        "W",
    ],
    "Color": [
        "Black",
        "Blue",
        "Green",
        "Red",
        "White",
    ],
    "Layout": [
        "normal",
    ],
    "Format": [
        "Brawl",
        "Commander",
        "Legacy",
        "Modern",
        "Standard",
        "Vintage",
    ],
    "LegalityEnum": [
        "Banned",
        "Legal",
        "Restricted",
    ],
    "Rarity": [
        "Basic Land",
        "Common",
        "Rare",
        "Uncommon",
    ],
    "Supertype": [
        "Basic",
    ],
    "Type": [
        "Artifact",
        "Creature",
        "Enchantment",
        "Instant",
        "Land",
        "Sorcery",
    ],
};

module.exports = {
    "lEAExtrasToJson": lEAExtrasToJson,
    "toLEAExtras": toLEAExtras,
};
